egs/wsj/s5/utils/prepare_extended_lang.sh

```diff
@@ -0,0 +1,182 @@
+#!/bin/bash
+# Copyright 2018  Xiaohui Zhang
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+# WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+# MERCHANTABLITY OR NON-INFRINGEMENT.
+# See the Apache 2 License for the specific language governing permissions and
+# limitations under the License.
+
+# This script adds word-position-dependent phones and constructs a host of other
+# derived files, that go in data/lang/.
+
+# Begin configuration section.
+oov_unigram_prob=0.0001 # Unigram LM probability we'll assign to each word outside the original vocab.
+srilm_opts="-subset -prune-lowprobs -unk"
+prep_lang_opts=
+lm=
+to_carpa=false # convert the language model to const arpa format rather than FST 
+stage=0
+# end configuration sections
+
+echo "$0 $@"  # Print the command line for logging
+
+. utils/parse_options.sh
+
+if [ $# -ne 7 ]; then
+  echo "usage: utils/prepare_extended_lang.sh <dict-src-dir> <oov-dict-entry> <extra-lexicon> <phone-symbol-table> <extended-dict-dir> <tmp-dir> <extended-lang-dir>"
+  echo "e.g.: utils/prepare_extended_lang.sh data/local/dict <SPOKEN_NOISE> lexicon_extra.txt data/lang/phones.txt data/local/dict_ext data/local/lang_ext data/lang_ext"
+  echo "The goal is to extend the lexicon from <dict-src-dir> with extra lexical entries from <extra-lexicon>, putting the extended lexicon into <extended-dict-dir>,"
+  echo "and then build a valid lang dir <extended-lang-dir>."
+  echo "<dict-src-dir> must be a valid dictionary dir and <oov-dict-entry> is the oov word (see utils/prepare_lang.sh for details)."
+  echo "If a language model is provided, we'll also extend the language model and re-build G.fst (or G.carpa), by inserting extra words as extra unigrams,"
+  echo "with probalities as the same as the unigram prob. (optionally boosted) of <oov-dict-entry> from the original LM."
```

Comment:
It's not clear what 'optionally boosted' here means-- and I don't think it would be appropriate to give the OOV-word prob to each individual OOV word, as there could be many of them.

Reply:
oh that's was the old comment created when I used a scaled version os the <unk>'s lm unigram prob before. I'll change it. BTW, when you said we shouldn't give the OOV-word prob to each individual OOV word, did you mean we should allow different probs for each OOV word? I think it's exactly what this script does: https://github.com/kaldi-asr/kaldi/blob/master/egs/babel/s5d/local/arpa2G.sh Actually, I just realized I, should probably remove the part dealing with expanding G.fst in my script and just use the above script for this purpose. Also I think we should move arpa2G.sh to utils level, e.g. utils/format_lm_extended.sh or even just modify utils/format_lm.sh to allow adding OOVs. What do you think? @jtrmal your opinion is also welcomed (basically I'm creating scripts for people to add words into L and G in test time).

Reply:
What I meant was that, regardless of whether the individual probs of the previously-unknown words are all the same or not, they should likely be smaller than the 'unk' class.  However, we have to be a bit careful with dividing the prob of the 'unk' class in the LM, between the previously-OOV words.  Different estimation formulas give very different values for the unk prob, that may or may not be appropriate here (and this depends on how the vocab and the data were prepared too),  so it might well be necessary to allow the user to manually override the unk-prob from the LM somehow, in case the value in the ARPA was unsuitable for some reason.

I do think that separating the 'G' parts of it from the rest of the 'lang' preparation is a good idea.  However, arpa2G.sh is not a good starting point.  That is a rather outdated script; these days, we do it more efficiently from the C++, as you can see from egs/ami/s5b/local/ami_format_data.sh:
```
gunzip -c "$arpa_lm" | \
  arpa2fst --disambig-symbol=#0 \
           --read-symbol-table=data/lang_test/words.txt - data/lang_test/G.fst
```
So no script in utils/ is needed just to format the LM.
I think what might be cleaner is to have some kind of script that would modify the ARPA file as part of a pipe, so you could do
```
gunzip -c "$arpa_lm" | \
  some_command_that_messes_with_unk_probs | \
  arpa2fst --disambig-symbol=#0 \
           --read-symbol-table=data/lang_test/words.txt - data/lang_test/G.fst
```


I think it probably would be a good idea to

Reply:
Thanks! I agree. Regarding overriding the <unk> prob. I think it's an independent issue. As we discussed before we might not want to consider the prob of <unk> class in LM when adding new words, since we don't know how it's estimated. But we should allow the user to change it somewhere else. For now I'm doing this by scaling all arcs with output symbol <unk> in HCLG. Do you think we want to commit such a script under utils? 

Sure I think it's a good idea to create a script (Did you mean some local commands or a script to put under utils? I suggest a script since it might be long) to add OOV & probs to an ARPA file. I think the mechanism from arpa2G.sh might still be good: ask the user the provide an oov2prob file, and an --unk-fraction scalar scaling the OOV probs. These probs will be then ceilinged by the smallest unigram prob existing in the ARPA. What do you think? 

At the end, you mentioned "I think it probably would be a good idea to". to what?

Reply:
In `some_command_that_messes_with_unk_probs`, my intention was that it was some kind of script under utils/lang/.  I don't think applying a ceiling based on the smallest unigram prob in the ARPA is a good idea, as there are situations where, due to domain mismatch, the new OOV words might be quite probable.  For the rest of the design-- I suggest you come up with something, and we discuss afterward.
I don't know what I was going to say with 'I think it probably would be a good idea to'-- forget that.

Reply:
done. I also added a perl script for inserting extra unigrams into ARPA lm.